### ISSUE TITLE
Setting Jackson Timezone Doesn't Work for ZonedDateTime Instances

### ISSUE THREAD

**Issue by tunix** [spring-projects/spring-boot#5513]
**Spring Boot Version:** 1.3.3
**Java Version:** 1.8.0_77

I've a 3rd party (XML based) web service which returns a date along with other information. I need to change the timezone it's sent (let's say from GMT+6 to GMT+2) while returning the response from the controller. However it's shown in the original timezone despite I set `spring.jackson.time-zone` to `GMT+2`. I've also tried the setter methods of Jackson2ObjectMapperBuilder but that didn't work either.

In my original project, when I read the response (via RestTemplate) from the 3rd party service it's automatically converted into UTC and I cannot set the timezone to whatever I want. I don't know why but it's not converted to UTC in the sample code I'm sharing.

**Comment by philwebb** [comment]
It appears that the `timeZone` and `dateFormat` Jackson settings only apply to `java.util.Date`. Take a look at this [stack overflow](http://stackoverflow.com/questions/29571587/deserializing-localdatetime-with-jackson-jsr310-module) answer for details.

If we want to apply the date format to JSR-310 it appears we'll need to re-register a bunch of JSR-310 converters directly ourselves.

For now I'd suggest trying to create your own `Module` to set the date formatter.

**Comment by philwebb** [comment]
@wilkinsona Do you think it's worth us trying to support JSR 310 in the same way as we do Joda?

**Comment by snicoll** [comment]
I thought all that was required was adding

```
<dependency>
  <groupId>com.fasterxml.jackson.datatype</groupId>
  <artifactId>jackson-datatype-jsr310</artifactId>
</dependency>
```

See https://github.com/snicoll-scratches/modern-component-design-spring-4.2/

**Comment by wilkinsona** [comment]
@tunix Thanks for the reproduction. Unfortunately, it has left me confused.

In `application.properties` you have explicitly told Jackson not to perform timezone conversion during deserialisation:

```
spring.jackson.deserialization.adjust_dates_to_context_time_zone=false
```

This means that `InstantDeserializer` skips [the logic](https://github.com/FasterXML/jackson-datatype-jsr310/blob/jackson-datatype-jsr310-2.6.5/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java#L153-L155) that would change the timezone.

If I update your `application.properties` to enable time zone adjustment the time in the response is shifted into GMT+2:

```
$ curl localhost:8080/test
{"createdAt":"2016-03-27T20:15:45+02:00"}[
```

In short, as far as I can tell, this is working fine. What have I missed?

**Comment by tunix** [comment]
@wilkinsona thanks for the detailed answer! I thought `spring.jackson.time-zone` would affect both deserialization & serialization. If you set `spring.jackson.deserialization.adjust_dates_to_context_time_zone` to false again, you can see that `spring.jackson.time-zone` has no effect on serialization.

In my case, a use RestTemplate to fetch the XML data. Jackson converts that data into objects but somehow it's converted to UTC during deserialization and I can't change the serialization format whatsoever. I couldn't find where it gets converted to UTC so I'm unable to solve this issue.

**Comment by wilkinsona** [comment]
@tunix `RestTemplate` doesn't appear to be used anywhere in the reproduction that you provided. It's also worth noting that if you create your own `RestTemplate` it'll used its own `ObjectMapper` by default. Are you customizing your `RestTemplate` with a converter that uses Boot's auto-configured `ObjectMapper`? Perhaps you could update your reproduction (another PR against `spring-boot-issues` would be ideal) so that it includes some usage of `RestTemplate` that fails due to the timezone being incorrect?

**Comment by tunix** [comment]
@wilkinsona Yes I've a subclass of RestTemplate which terminates SSL cert. verification and uses basic auth. It's probably using its own ObjectMapper so I'll try to use Spring's ObjectMapper instead. Because when I remove `@JsonFormat` annotations from all LocalDate's and ZonedDateTime's it cannot convert XML data into objects. I should have understood that this would be related to my RestTemplate subclass having its own ObjectMapper instance.

The API I'm consuming is private. I'll try to create a mock API so that I can push another PR including my RestTemplate subclass. Thanks for the pointers!

**Comment by tunix** [comment]
I've added a mock API with a custom RestTemplate. Now I get:

```

**Comment by wilkinsona** [comment]
@tunix Is that good or bad? I could probably figure it out, but a test with an assertion that fails would be a much better way of illustrating the output that you're getting and the output that you expect to get.

Looking at your new PR, I can't see where you're configuring your custom `RestTemplate` to use the auto-configured `ObjectMapper`. Perhaps I missed it? If not, you'll need to configure it with a `MappingJackson2XmlHttpMessageConverter` that itself has been configured with the context's `ObjectMapper`.

**Comment by tunix** [comment]
@wilkinsona I expected to see the whole time zone like this:

```
{"createdAt":"2016-03-27T18:15:45+06:00"}
```

I've added a test for my expectation. I tried autowiring MappingJackson2XmlHttpMessageConverter since Spring globally defines one but the result didn't change.

**Comment by tunix** [comment]
With latest changes to my PR (https://github.com/spring-projects/spring-boot-issues/pull/46) it works.
Thanks for all your help! I believe we can close this ticket.

**Comment by wilkinsona** [comment]
Excellent

**Comment by tunix** [comment]
Hi @wilkinsona ,

If I remove `@JsonFormat(pattern...)` from `Transaction.java` then deserialization of `2016-03-28T00:15:45.000+0600` fails with error:

```

**Comment by wilkinsona** [comment]
@tunix That's really a separate issue to the timezone handling that we discussed above. As @philwebb said above, the object mapper's date format (which is what `spring.jackson.date-format` configures) only applies to `java.util.Date` instances. If you want to use a non-ISO-8601 format with `java.time` types, then you'll need to register your own deserialisers configured with an appropriate `DateTimeFormatter`. If you'd like us to see if Spring Boot could do that for you please open a new issue.
